File: zig/src/poller.zig
const std = @import("std");
const config_mod = @import("config.zig");
const db_mod = @import("db.zig");
const http_client = @import("http_client.zig");

const checkpoint_interval_polls = 10;

pub const HealthStatus = enum {
    unknown,
    ok,
    err,
};

pub const DeviceHealth = struct {
    ip: []const u8,
    label: []const u8,
    status: HealthStatus,
    last_success: i64,
    last_error: i64,
    last_error_message: [256]u8,
    last_error_message_len: usize,
    consecutive_failures: u32,

    pub fn init(ip: []const u8, label: []const u8) DeviceHealth {
        return DeviceHealth{
            .ip = ip,
            .label = label,
            .status = .unknown,
            .last_success = 0,
            .last_error = 0,
            .last_error_message = [_]u8{0} ** 256,
            .last_error_message_len = 0,
            .consecutive_failures = 0,
        };
    }

    fn recordSuccess(self: *DeviceHealth) void {
        self.status = .ok;
        self.last_success = db_mod.nowMillis();
        self.last_error_message_len = 0;
        self.consecutive_failures = 0;
    }

    fn recordFailure(self: *DeviceHealth, msg: []const u8) void {
        self.status = .err;
        self.last_error = db_mod.nowMillis();
        const copy_len = @min(msg.len, self.last_error_message.len);
        @memcpy(self.last_error_message[0..copy_len], msg[0..copy_len]);
        self.last_error_message_len = copy_len;
        self.consecutive_failures +|= 1; // saturating add — no overflow
    }

    fn errorMessage(self: *const DeviceHealth) []const u8 {
        return self.last_error_message[0..self.last_error_message_len];
    }
};

pub const PollerState = struct {
    db: db_mod.SqliteDb,
    db_mutex: *std.Thread.Mutex,
    health: []DeviceHealth,
    health_mutex: *std.Thread.Mutex,
    config: *const config_mod.Config,
    shutdown: *std.atomic.Value(bool),
    poll_successes: std.atomic.Value(u64) = std.atomic.Value(u64).init(0),
    poll_failures: std.atomic.Value(u64) = std.atomic.Value(u64).init(0),
};

fn formatOptF64(buf: []u8, val: ?f64) []const u8 {
    if (val) |v| {
        return std.fmt.bufPrint(buf, "{d:.2}", .{v}) catch "?";
    }
    return "N/A";
}

fn formatOptI64(buf: []u8, val: ?i64) []const u8 {
    if (val) |v| {
        return std.fmt.bufPrint(buf, "{d}", .{v}) catch "?";
    }
    return "N/A";
}

fn fetchDevice(state: *PollerState, idx: usize) void {
    const ip = state.config.devices[idx].ip;
    const label = state.config.devices[idx].label;

    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const body = http_client.httpGet(allocator, ip, "/measures/current", state.config.fetch_timeout_ms) catch |e| {
        var err_buf: [256]u8 = [_]u8{0} ** 256;
        const err_msg = std.fmt.bufPrint(&err_buf, "fetch failed: {s}", .{@errorName(e)}) catch "fetch failed";
        std.log.err("[poller] {s} ({s}): {s}", .{ label, ip, err_msg });
        _ = state.poll_failures.fetchAdd(1, .monotonic);
        state.health_mutex.lock();
        defer state.health_mutex.unlock();
        state.health[idx].recordFailure(err_msg);
        return;
    };

    const parsed = std.json.parseFromSlice(std.json.Value, allocator, body, .{}) catch {
        const msg = "JSON parse error";
        std.log.err("[poller] {s} ({s}): {s}", .{ label, ip, msg });
        _ = state.poll_failures.fetchAdd(1, .monotonic);
        state.health_mutex.lock();
        defer state.health_mutex.unlock();
        state.health[idx].recordFailure(msg);
        return;
    };
    const data = parsed.value;

    if (data != .object) {
        const msg = "unexpected response type: not an object";
        std.log.err("[poller] {s} ({s}): {s}", .{ label, ip, msg });
        _ = state.poll_failures.fetchAdd(1, .monotonic);
        state.health_mutex.lock();
        defer state.health_mutex.unlock();
        state.health[idx].recordFailure(msg);
        return;
    }

    // Insert into DB
    {
        state.db_mutex.lock();
        defer state.db_mutex.unlock();
        db_mod.insertReading(state.db, ip, data, body) catch {
            const msg = "DB insert failed";
            std.log.err("[poller] {s} ({s}): {s}", .{ label, ip, msg });
            _ = state.poll_failures.fetchAdd(1, .monotonic);
            state.health_mutex.lock();
            defer state.health_mutex.unlock();
            state.health[idx].recordFailure(msg);
            return;
        };
    }

    // Extract values for log line (safe — returns null for NaN/Inf)
    const pm02 = extractOptF64(data, "pm02");
    const rco2 = extractOptI64(data, "rco2");
    const atmp = extractOptF64(data, "atmp");

    var pm02_buf: [32]u8 = [_]u8{0} ** 32;
    var rco2_buf: [32]u8 = [_]u8{0} ** 32;
    var atmp_buf: [32]u8 = [_]u8{0} ** 32;
    const pm02_s = formatOptF64(&pm02_buf, pm02);
    const rco2_s = formatOptI64(&rco2_buf, rco2);
    const atmp_s = formatOptF64(&atmp_buf, atmp);

    _ = state.poll_successes.fetchAdd(1, .monotonic);

    {
        state.health_mutex.lock();
        defer state.health_mutex.unlock();
        state.health[idx].recordSuccess();
    }

    std.log.info("[poller] {s} ({s}): OK — PM2.5={s}, CO2={s}, T={s}°C", .{ label, ip, pm02_s, rco2_s, atmp_s });
}

fn extractOptF64(data: std.json.Value, key: []const u8) ?f64 {
    const val = data.object.get(key) orelse return null;
    return switch (val) {
        .float => |f| if (std.math.isNan(f) or std.math.isInf(f)) null else f,
        .integer => |i| @as(f64, @floatFromInt(i)),
        else => null,
    };
}

fn extractOptI64(data: std.json.Value, key: []const u8) ?i64 {
    const val = data.object.get(key) orelse return null;
    return switch (val) {
        .integer => |i| i,
        .float => |f| blk: {
            if (std.math.isNan(f) or std.math.isInf(f)) break :blk null;
            if (f > @as(f64, @floatFromInt(std.math.maxInt(i64)))) break :blk null;
            if (f < @as(f64, @floatFromInt(std.math.minInt(i64)))) break :blk null;
            break :blk @as(i64, @intFromFloat(f));
        },
        else => null,
    };
}

fn pollAll(state: *PollerState) void {
    for (0..state.config.device_count) |i| {
        fetchDevice(state, i);
    }
}

pub fn run(state: *PollerState) void {
    std.log.info("[poller] Starting — polling {d} devices every {d}s", .{
        state.config.device_count,
        state.config.poll_interval_ms / 1000,
    });

    // Initialize health
    {
        state.health_mutex.lock();
        defer state.health_mutex.unlock();
        for (0..state.config.device_count) |i| {
            state.health[i] = DeviceHealth.init(
                state.config.devices[i].ip,
                state.config.devices[i].label,
            );
        }
    }

    // Initial poll
    pollAll(state);

    var poll_count: u32 = 0;
    const sleep_ns: u64 = @as(u64, state.config.poll_interval_ms) * 1_000_000;

    while (true) {
        std.Thread.sleep(sleep_ns);

        if (state.shutdown.load(.acquire)) {
            std.log.info("[poller] Stopped", .{});
            break;
        }

        pollAll(state);
        poll_count +|= 1; // saturating

        if (poll_count % checkpoint_interval_polls == 0) {
            state.db_mutex.lock();
            defer state.db_mutex.unlock();
            db_mod.checkpoint(state.db) catch {
                std.log.err("[poller] WAL checkpoint failed", .{});
            };
        }
    }
}

pub fn getHealthJson(allocator: std.mem.Allocator, health: []const DeviceHealth, count: usize, health_mutex: *std.Thread.Mutex) !std.json.Value {
    var arr = std.json.Array.init(allocator);

    health_mutex.lock();
    defer health_mutex.unlock();

    for (0..count) |i| {
        const h = &health[i];
        var obj = std.json.ObjectMap.init(allocator);

        try obj.put("ip", .{ .string = h.ip });
        try obj.put("label", .{ .string = h.label });

        const status_str: []const u8 = switch (h.status) {
            .unknown => "unknown",
            .ok => "ok",
            .err => "error",
        };
        try obj.put("status", .{ .string = status_str });

        try obj.put("lastSuccess", if (h.last_success != 0) .{ .integer = h.last_success } else .null);
        try obj.put("lastError", if (h.last_error != 0) .{ .integer = h.last_error } else .null);

        if (h.last_error_message_len > 0) {
            const msg = try allocator.dupe(u8, h.errorMessage());
            try obj.put("lastErrorMessage", .{ .string = msg });
        } else {
            try obj.put("lastErrorMessage", .null);
        }

        try obj.put("consecutiveFailures", .{ .integer = @as(i64, h.consecutive_failures) });

        try arr.append(.{ .object = obj });
    }

    return .{ .array = arr };
}

File: c/src/poller.c
#include "poller.h"
#include "app.h"
#include "db.h"
#include "http_client.h"
#include "json.h"

#include <stdatomic.h>
#include <stdio.h>
#include <string.h>
#include <time.h>

#define CHECKPOINT_INTERVAL_POLLS 10

static atomic_uint_fast64_t poll_successes = 0;
static atomic_uint_fast64_t poll_failures  = 0;

static void record_success(DeviceHealth *h)
{
    h->status = HEALTH_OK;
    h->last_success = db_now_millis();
    h->last_error_message[0] = '\0';
    h->consecutive_failures = 0;
}

static void record_failure(DeviceHealth *h, const char *msg)
{
    h->status = HEALTH_ERROR;
    h->last_error = db_now_millis();
    snprintf(h->last_error_message, sizeof(h->last_error_message), "%s", msg);
    h->consecutive_failures++;
}

static void format_optional_f64(char *buf, size_t size, bool has, double val)
{
    if (has) {
        snprintf(buf, size, "%.2f", val);
    } else {
        snprintf(buf, size, "N/A");
    }
}

static void format_optional_i64(char *buf, size_t size, bool has, int64_t val)
{
    if (has) {
        snprintf(buf, size, "%lld", (long long)val);
    } else {
        snprintf(buf, size, "N/A");
    }
}

static void fetch_device(AppState *state, size_t idx)
{
    const char *ip    = state->config.devices[idx].ip;
    const char *label = state->config.devices[idx].label;

    char errbuf[256];
    errbuf[0] = '\0';
    char *body = http_get(ip, "/measures/current", state->config.fetch_timeout_ms,
                          errbuf, sizeof(errbuf));

    if (!body) {
        fprintf(stderr, "[poller] %s (%s): fetch failed: %s\n", label, ip, errbuf);
        atomic_fetch_add(&poll_failures, 1);
        pthread_rwlock_wrlock(&state->health_lock);
        record_failure(&state->health[idx], errbuf);
        pthread_rwlock_unlock(&state->health_lock);
        return;
    }

    JsonError jerr;
    JsonValue *data = json_parse(body, strlen(body), &jerr);
    free(body);

    if (!data) {
        char msg[256];
        snprintf(msg, sizeof(msg), "JSON parse error at pos %zu", jerr.pos);
        fprintf(stderr, "[poller] %s (%s): %s\n", label, ip, msg);
        atomic_fetch_add(&poll_failures, 1);
        pthread_rwlock_wrlock(&state->health_lock);
        record_failure(&state->health[idx], msg);
        pthread_rwlock_unlock(&state->health_lock);
        return;
    }

    if (!json_is_object(data)) {
        const char *msg = "unexpected response type: not an object";
        fprintf(stderr, "[poller] %s (%s): %s\n", label, ip, msg);
        atomic_fetch_add(&poll_failures, 1);
        pthread_rwlock_wrlock(&state->health_lock);
        record_failure(&state->health[idx], msg);
        pthread_rwlock_unlock(&state->health_lock);
        json_free(data);
        return;
    }

    /* Insert into DB under mutex */
    pthread_mutex_lock(&state->db_mutex);
    int rc = db_insert_reading(state->db, ip, data);
    pthread_mutex_unlock(&state->db_mutex);

    if (rc != 0) {
        const char *msg = "DB insert failed";
        fprintf(stderr, "[poller] %s (%s): %s\n", label, ip, msg);
        atomic_fetch_add(&poll_failures, 1);
        pthread_rwlock_wrlock(&state->health_lock);
        record_failure(&state->health[idx], msg);
        pthread_rwlock_unlock(&state->health_lock);
        json_free(data);
        return;
    }

    /* Log success */
    bool ok;
    double pm02 = json_as_f64(json_get(data, "pm02"), &ok);
    bool has_pm02 = ok;
    int64_t rco2 = json_as_i64(json_get(data, "rco2"), &ok);
    bool has_rco2 = ok;
    double atmp = json_as_f64(json_get(data, "atmp"), &ok);
    bool has_atmp = ok;

    char pm02_s[32], rco2_s[32], atmp_s[32];
    format_optional_f64(pm02_s, sizeof(pm02_s), has_pm02, pm02);
    format_optional_i64(rco2_s, sizeof(rco2_s), has_rco2, rco2);
    format_optional_f64(atmp_s, sizeof(atmp_s), has_atmp, atmp);

    atomic_fetch_add(&poll_successes, 1);
    pthread_rwlock_wrlock(&state->health_lock);
    record_success(&state->health[idx]);
    pthread_rwlock_unlock(&state->health_lock);

    fprintf(stderr, "[poller] %s (%s): OK — PM2.5=%s, CO2=%s, T=%s°C\n",
            label, ip, pm02_s, rco2_s, atmp_s);

    json_free(data);
}

static void poll_all(AppState *state)
{
    for (size_t i = 0; i < state->config.device_count; i++) {
        fetch_device(state, i);
    }
}

void *poller_run(void *arg)
{
    AppState *state = (AppState *)arg;

    fprintf(stderr, "[poller] Starting — polling %zu devices every %us\n",
            state->config.device_count, state->config.poll_interval_ms / 1000u);

    /* Init health */
    pthread_rwlock_wrlock(&state->health_lock);
    for (size_t i = 0; i < state->config.device_count; i++) {
        state->health[i].ip    = state->config.devices[i].ip;
        state->health[i].label = state->config.devices[i].label;
        state->health[i].status = HEALTH_UNKNOWN;
        state->health[i].last_success = 0;
        state->health[i].last_error = 0;
        state->health[i].last_error_message[0] = '\0';
        state->health[i].consecutive_failures = 0;
    }
    pthread_rwlock_unlock(&state->health_lock);

    /* Initial poll */
    poll_all(state);

    uint32_t poll_count = 0;

    struct timespec sleep_ts;
    sleep_ts.tv_sec  = (time_t)(state->config.poll_interval_ms / 1000u);
    sleep_ts.tv_nsec = (long)((state->config.poll_interval_ms % 1000u) * 1000000u);

    for (;;) {
        nanosleep(&sleep_ts, NULL);

        if (atomic_load(&state->shutdown)) {
            fprintf(stderr, "[poller] Stopped\n");
            break;
        }

        poll_all(state);
        poll_count++;

        if (poll_count % CHECKPOINT_INTERVAL_POLLS == 0) {
            pthread_mutex_lock(&state->db_mutex);
            if (db_checkpoint(state->db) != 0) {
                fprintf(stderr, "[poller] WAL checkpoint failed\n");
            }
            pthread_mutex_unlock(&state->db_mutex);
        }
    }

    return NULL;
}

/* ---- poll counters ---- */

void poller_get_counters(uint64_t *successes, uint64_t *failures)
{
    *successes = atomic_load(&poll_successes);
    *failures  = atomic_load(&poll_failures);
}

/* ---- health JSON ---- */

JsonValue *poller_get_health_json(struct AppState *state)
{
    JsonValue *arr = json_array_new();

    pthread_rwlock_rdlock(&state->health_lock);
    for (size_t i = 0; i < state->config.device_count; i++) {
        const DeviceHealth *h = &state->health[i];

        JsonValue *obj = json_object_new();
        json_object_set(obj, "ip",    json_string(h->ip));
        json_object_set(obj, "label", json_string(h->label));

        const char *status_str = "unknown";
        if (h->status == HEALTH_OK)    status_str = "ok";
        if (h->status == HEALTH_ERROR) status_str = "error";
        json_object_set(obj, "status", json_string(status_str));

        json_object_set(obj, "lastSuccess",
            h->last_success ? json_from_i64(h->last_success) : json_null());
        json_object_set(obj, "lastError",
            h->last_error ? json_from_i64(h->last_error) : json_null());
        json_object_set(obj, "lastErrorMessage",
            h->last_error_message[0] ? json_string(h->last_error_message) : json_null());
        json_object_set(obj, "consecutiveFailures",
            json_number((double)h->consecutive_failures));

        json_array_push(arr, obj);
    }
    pthread_rwlock_unlock(&state->health_lock);

    return arr;
}
